subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,588 @@
+		out->out_cpkt->offset += payload_len;
+	} else {
+		payload_len = -ENOMEM;
+	}
+
+	return payload_len;
+}
+
+static size_t put_bool(struct lwm2m_output_context *out, struct lwm2m_obj_path *path, bool value)
+{
+	cbor_state_t states[1];
+	int payload_len;
+
+	new_state(states, 1, CPKT_BUF_W_REGION(out->out_cpkt), 1);
+
+	bool ret = boolx_encode(states, &value);
+
+	if (ret) {
+		payload_len = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+		out->out_cpkt->offset += payload_len;
+	} else {
+		payload_len = -ENOMEM;
+	}
+
+	return payload_len;
+}
+
+static size_t put_objlnk(struct lwm2m_output_context *out, struct lwm2m_obj_path *path,
+			 struct lwm2m_objlnk *value)
+{
+	char objlnk[sizeof("65535:65535")] = { 0 };
+
+	snprintk(objlnk, sizeof(objlnk), "%" PRIu16 ":%" PRIu16 "", value->obj_id, value->obj_inst);
+
+	return put_string(out, path, objlnk, strlen(objlnk) + 1);
+}
+
+static size_t get_s64(struct lwm2m_input_context *in, int64_t *value)
+{
+	cbor_state_t states[1];
+	/* TODO: CBOR data item header + the variable
+	 * 1 + 8 once 64-bit values are supported, 1 + 4 for now for uint32s.
+	 */
+	new_state(states, 1, ICTX_BUF_R_PTR(in), MIN(5, ICTX_BUF_R_LEFT_SZ(in)), 1);
+
+	if (!uintx32_decode(states, (uint32_t *)value)) {
+		LOG_WRN("unable to decode a 64-bit(uint32) integer value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	/* TODO: length of the original binary data or the decoded value? */
+	return len;
+}
+
+static size_t get_s32(struct lwm2m_input_context *in, int32_t *value)
+{
+	cbor_state_t states[1];
+
+	new_state(states, 1, ICTX_BUF_R_PTR(in), MIN(5, ICTX_BUF_R_LEFT_SZ(in)), 1);
+
+	if (!intx32_decode(states, value)) {
+		LOG_WRN("unable to decode a 32-bit integer value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static size_t get_string(struct lwm2m_input_context *in, uint8_t *value, size_t buflen)
+{
+	cbor_state_t states[1];
+	cbor_string_type_t hndl = { .value = value, .len = buflen };
+
+	new_state(states, 1, ICTX_BUF_R_REGION(in), 1);
+
+	if (!tstrx_decode(states, &hndl)) {
+		LOG_WRN("unable to decode a string");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+/* Gets time decoded as a numeric string.
+ *
+ * return 0 on success, -EBADMSG if decoding fails or -EFAIL if value is invalid
+ */
+static int get_time_string(struct lwm2m_input_context *in, int64_t *value)
+{
+	cbor_state_t states[1];
+	char time_str[sizeof("4294967295")] = { 0 };
+	cbor_string_type_t hndl = { .value = time_str, .len = sizeof(time_str) - 1 };
+
+	new_state(states, 1, ICTX_BUF_R_REGION(in), 1);
+
+	if (!tstrx_decode(states, &hndl)) {
+		return -EBADMSG;
+	}
+
+	/* TODO: decode date/time string */
+	LOG_DBG("decoding a date/time string not supported");
+
+	return -ENOTSUP;
```

Comment:
I realized that the date/time strings are not supported yet. Leshan passes time in a numerical format so this hasn't been an issue yet.